ensure reflection returns all names in a module

vtjnash opened the pull request:
fix #17140

vtjnash committed:
ensure reflection returns all names in a module

fix #17140

tkelman commented:
Where are the ASCIIString errors coming from here?

vtjnash commented:
we're running with `--depwarn=error`, and this changes reflection to also return deprecated bindings

tkelman commented:
Hm. So do we temporarily swallow warnings any time we use `all=true` in the tests then?

vtjnash commented:
no, I'm working on a more complete fix. But so far, fixing this has uncovered 3 bugs in our testing :/ and caused me to fix examples/typetree.jl to be prettier :). Also it should save a small amount of memory.

vtjnash committed:
fix ambiguity test

now that names(all=true) returns all names,
can fix the ambiguity test to check all MethodTables

vtjnash committed:
reuse some datatype layouts

eventually we may want to hash-cons all of these.
in the meantime, can avoid copying the layout struct
for common known types

vtjnash committed:
fix and improve typetree.jl example

make it recursive over all modules,
update deprecations,
format output a bit better (removing duplicates better)

vtjnash committed:
fix dumptypes

this method was broken and extremely slow
make it more correct and add tests

vtjnash committed:
fix a numbers tests to be non-empty

subtypes(Complex) is the empty set since 7f9325cf improved it

vtjnash committed:
improve testing coverage and implementation of dump

tkelman commented on test/show.jl:
```diff
@@ -278,20 +278,19 @@ end
     local rdout, wrout, rderr, wrerr, out, err, rd, wr
     try
-        rd, wr = redirect_stdout()
-        @test dump(STDERR) == nothing
```
why are you deleting this?

vtjnash replied:
it wasn't checking it's output so it's just artificially inflating the coverage. I added more than enough tests below of the same to make up for it.

vtjnash commented:
Test failure seems to have been the libopenspecfun mod2pi issue, so ready to merge now

StefanKarpinski commented:
This failure appears to be something else: https://travis-ci.org/JuliaLang/julia/jobs/145942953. If you're sure that it's unrelated, go ahead and merge – I can't really tell.

vtjnash commented:
@yuyichao is it possible that the size of the `.eh_frame` section is not exactly equal to the end of the last symbol in it?

yuyichao commented:
I don't know. @Keno 

In any case, I think we should stop processing after we hit a terminator? So maybe

```diff
diff --git a/src/debuginfo.cpp b/src/debuginfo.cpp
index 4901eab..5a6c531 100644
--- a/src/debuginfo.cpp
+++ b/src/debuginfo.cpp
@@ -1357,7 +1357,7 @@ extern "C" void __register_frame(void*);
 extern "C" void __deregister_frame(void*);
 
 template <typename callback>
-static const char *processFDE(const char *Entry, callback f)
+static uint32_t processFDE(const char *Entry, callback f)
 {
     const char *P = Entry;
     uint32_t Length = *((const uint32_t *)P);
@@ -1368,7 +1368,7 @@ static const char *processFDE(const char *Entry, callback f)
     if (Offset != 0 && Length != 0) {
         f(Entry);
     }
-    return P + Length;
+    return Length;
 }
 
 template <typename callback>
@@ -1376,9 +1376,12 @@ static void processFDEs(const char *EHFrameAddr, size_t EHFrameSize, callback f)
 {
     const char *P = (const char*)EHFrameAddr;
     const char *End = P + EHFrameSize;
-    do  {
-        P = processFDE(P, f);
-    } while(P != End);
+    do {
+        uint32_t Length = processFDE(P, f);
+        if (Length == 0)
+            break;
+        P += Length + 4;
+    } while (P != End);
 }
 #endif
 ```

JeffBezanson commented on base/interactiveutil.jl:
```diff
@@ -670,7 +670,7 @@ function summarysize(obj::Module, seen, excl)
     size::Int = Core.sizeof(obj)
     for binding in names(obj, true)
-        if isdefined(obj, binding)
+        if isdefined(obj, binding) && !isdeprecated(obj, binding)
```
This seems like a strange place to exclude deprecated items.

vtjnash replied:
agreed. but we need some non-warning form of `getfield` so that we can reflect on them without causing warnings to be printed (and with `--depwarn=error` to fail). However, I'm not sure what that API should look

maleadt commented:
This makes `testall` fail on my system because of the new tests for `dumptypes` introduced in 1b470f5a0a833903b7c724c618f9e2b92ccd5caa, more specifically the `@test startswith(repr, "Any\n  Base.")`.

When running in isolation, `sprint(dump, Any)` starts with:
```
Any
  Base.=> = Pair{A,B}
  Base.AbstractCartesianIndex{N}
    Base.CartesianIndex = CartesianIndex{N}
    CartesianIndex{N}
  AbstractChannel
    Channel{T}
...
```

However, when running `make testall`, it starts with:
```
Any
  Main.#1181#rettype = Tuple{Int64,Int64}
  Main.#1185#rettype = Tuple{Int64,Int64,Int64}
  Base.=> = Pair{A,B}
  Base.AbstractCartesianIndex{N}
    Base.CartesianIndex = CartesianIndex{N}
    CartesianIndex{N}
  AbstractChannel
    Channel{T}
...
```

vtjnash commented:
alright, I added that as an option for that test too: https://github.com/JuliaLang/julia/commit/8a9fd10b0a317b952575e9dd230b901342636f05
